### DISCUSSION TITLE
Best practice when using an async database session but needing to execute database queries in a synchronous task

### DISCUSSION THREAD

**Discussion by ThomasAitken** [fastapi/fastapi#10897] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def enrich_content(content_id: int):
    """
    Synchronous task that needs to interact with my DB.
    """

    # Fetch metadata from external source
    external_response = requests.get("external_url")
    ...
    # add database logic here


@router.post(
    "my_url",
)
async def edit_content(
    my_payload: PayloadSchema,
    async_db_session: DBSessionDep,
    background_tasks: BackgroundTasks,
):
    # asynchronous CRUD function
    content_id = await update_content(async_db_session, my_payload)
    background_tasks.add_task(enrich_content, content_id)
```

### Description

I am using SQLAlchemy in its async mode. So all my CRUD functions are executing in an async session using my database dependency (`DBSessionDep`). 

However, in one of my endpoints, I want to kick off a synchronous background task that fetches data from an external source and then uses that data to update one of my tables. My problem is that in the context of that task, I can't use my async session manager to execute these queries without doing something a bit unorthodox.

These are the 3 unorthodox solutions I can think of:

1. Create a new event loop inside the synchronous task in order to run the async database logic. This feels a bit wasteful and unconventional but I'm not sure the actual downsides.

2. Create an additional session manager class that creates a synchronous session just for use-cases such as this. This feels a bit clunky and has the downside of creating a second connection pool.

3. Still run the task in a separate thread but convert the whole thing to async, even though it has 0 performance benefit in this context (e.g. because I have to make an HTTP request).

Which of these is preferred overall? Or is there some option I've missed?

### Operating System

Linux

### FastAPI Version

0.103.2

### Pydantic Version

2.1.1

### Python Version

3.11.7

**Comment by ThomasAitken** [comment]
On further reflection, I have a few more thoughts that might assist other people if they happen to stumble across this.

First, one obvious solution I missed is using an async HTTP client library instead of requests and thereby making the entire task async. Assuming low scale, I suspect the performance of this task would be very similar to the synchronous version. I also suspect that this might be the recommendation of most FastAPI enthusiasts. That way, we keep things simple and consistent, even if it means that we have to do basically *everything* async.

But on the other hand, I was thinking that the ideal scenario for such a background task - and indeed most background tasks that I could imagine running - is that they run in a genuinely parallel way, i.e. in a separate process (note: FastAPI background tasks may run in a separate thread but will still be in the same process, so given the GIL, they will not genuinely be parallel, only concurrent). In particular, I think I will end up using Celery for my project. And it's easier and simpler to just do things synchronously in Celery. So for that reason I am leaning towards using a separate synchronous session manager and keeping this task fully synchronous so that it, and similar tasks I create, are ready for later porting to Celery.

With such an architecture, I may still want to use FastAPI's background tasks functionality but only for quite small tasks that I can easily keep async, and never for third-party API requests or heavy computation.

**Comment by aanchlia** [comment, marked as answer]
I think for true parallelism, using an async task queue like Celery is the way to go.
